src/address_space.rs

```diff
@@ -34,61 +57,156 @@ impl AddressSpace {
     pub fn new(name: &str) -> Self {
         Self {
             name: name.to_string(),
-            mappings: LinkedList::new(),
-        }
+            mappings: Vec::new(), // <- here I changed from LinkedList, for reasons
+        } // I encourage you to try other sparse representations - trees, DIY linked lists, ...
     }
 
     /// Add a mapping from a `DataSource` into this `AddressSpace`.
     ///
     /// # Errors
     /// If the desired mapping is invalid.
-    pub fn add_mapping<D: DataSource>(
-        &self,
-        source: &D,
+    /// TODO: how does our test in lib.rs succeed?
+    // pub fn add_mapping<'a, D: DataSource + 'a>(
+    //     &'a mut self,
+    pub fn add_mapping<D: DataSource + 'static>(
+        &mut self,
+        source: Arc<D>,
         offset: usize,
         span: usize,
+        flags: FlagBuilder,
     ) -> Result<VirtualAddress, &str> {
-        todo!()
+        let mut addr_iter = PAGE_SIZE; // let's not map page 0
+        let mut gap;
+        let adjusted_span: usize = span + PAGE_SIZE - (span % PAGE_SIZE);
+        for mapping in &self.mappings {
+            // space between end of last entry and start of this entry
+            gap = mapping.addr - addr_iter;
+            // we found a big free space!
+            // need to have a free page between each mapping
+            // for accidental overflows
+            if gap > adjusted_span + 2 * PAGE_SIZE {
+                break;
+            }
+            // move addr_iter to the next potentially free chunk of memory
+            addr_iter = mapping.addr + mapping.span;
+        }
+        // if we have enough space for our new mapping
+        if addr_iter + adjusted_span + 2 * PAGE_SIZE < VADDR_MAX {
+            // add free page between mappings
+            let mapping_addr = addr_iter + PAGE_SIZE;
+            let new_mapping = MapEntry::new(source, offset, adjusted_span, mapping_addr, flags);
+            self.mappings.push(new_mapping);
+            self.mappings.sort_by(|a, b| a.addr.cmp(&b.addr));
+            return Ok(mapping_addr);
+        }
+        Err("out of address space!")
     }
 
     /// Add a mapping from `DataSource` into this `AddressSpace` starting at a specific address.
     ///
     /// # Errors
     /// If there is insufficient room subsequent to `start`.
-    pub fn add_mapping_at<D: DataSource>(
-        &self,
-        source: &D,
+    pub fn add_mapping_at<D: DataSource + 'static>(
+        &mut self,
+        source: Arc<D>,
         offset: usize,
         span: usize,
         start: VirtualAddress,
+        flags: FlagBuilder,
     ) -> Result<(), &str> {
-        todo!()
+        let adjusted_span: usize = span + PAGE_SIZE - (span % PAGE_SIZE);
```

Comment:
The standard library provides the `next_multiple_of` method on integer types, which does this for you.